go/store/types/serial_message.go

```diff
@@ -53,125 +53,135 @@ func (sm SerialMessage) Hash(nbf *NomsBinFormat) (hash.Hash, error) {
 		hashes = msg.ParentAddrsBytes()
 		for i := 0; i < msg.ParentAddrsLength()/hash.ByteLen; i++ {
 			addr := hash.New(hashes[i*20 : (i+1)*20])
-			fmt.Fprintf(ret, "\t\t#%s\n", addr.String())
+			printWithIndendationLevel(level, ret, "\t\t#%s\n", addr.String())
 		}
-		fmt.Fprintf(ret, "\t}\n")
+		printWithIndendationLevel(level, ret, "\t}\n")
 
-		fmt.Fprintf(ret, "\tParentClosure: {\n")
+		printWithIndendationLevel(level, ret, "\tParentClosure: {\n")
 		hashes = msg.ParentClosureBytes()
 		for i := 0; i < msg.ParentClosureLength()/hash.ByteLen; i++ {
 			addr := hash.New(hashes[i*20 : (i+1)*20])
-			fmt.Fprintf(ret, "\t\t#%s\n", addr.String())
+			printWithIndendationLevel(level, ret, "\t\t#%s\n", addr.String())
 		}
-		fmt.Fprintf(ret, "\t}\n")
+		printWithIndendationLevel(level, ret, "\t}\n")
 
-		fmt.Fprintf(ret, "}")
+		printWithIndendationLevel(level, ret, "}")
 		return ret.String()
 	case serial.RootValueFileID:
 		msg := serial.GetRootAsRootValue(sm, serial.MessagePrefixSz)
 		ret := &strings.Builder{}
-		fmt.Fprintf(ret, "{\n")
-		fmt.Fprintf(ret, "\tFeatureVersion: %d\n", msg.FeatureVersion())
-		fmt.Fprintf(ret, "\tForeignKeys: #%s\n", hash.New(msg.ForeignKeyAddrBytes()).String())
-		fmt.Fprintf(ret, "\tTables: {\n\t%s", SerialMessage(msg.TablesBytes()).HumanReadableString())
-		fmt.Fprintf(ret, "\t}\n")
-		fmt.Fprintf(ret, "}")
+		printWithIndendationLevel(level, ret, "{\n")
+		printWithIndendationLevel(level, ret, "\tFeatureVersion: %d\n", msg.FeatureVersion())
+		printWithIndendationLevel(level, ret, "\tForeignKeys: #%s\n", hash.New(msg.ForeignKeyAddrBytes()).String())
+		printWithIndendationLevel(level, ret, "\tTables: %s\n",
+			SerialMessage(msg.TablesBytes()).humanReadableStringAtIndentationLevel(level+1))
+		printWithIndendationLevel(level, ret, "}")
 		return ret.String()
 	case serial.TableFileID:
 		msg := serial.GetRootAsTable(sm, serial.MessagePrefixSz)
 		ret := &strings.Builder{}
 
-		fmt.Fprintf(ret, "asdasdf {\n")
-		fmt.Fprintf(ret, "\tSchema: #%s\n", hash.New(msg.SchemaBytes()).String())
-		fmt.Fprintf(ret, "\tViolations: #%s\n", hash.New(msg.ViolationsBytes()).String())
+		printWithIndendationLevel(level, ret, "{\n")
+		printWithIndendationLevel(level, ret, "\tSchema: #%s\n", hash.New(msg.SchemaBytes()).String())
+		printWithIndendationLevel(level, ret, "\tViolations: #%s\n", hash.New(msg.ViolationsBytes()).String())
 		// TODO: merge conflicts, not stable yet
 
-		fmt.Fprintf(ret, "\tAutoinc: %d\n", msg.AutoIncrementValue())
-
-		// TODO: can't use tree package to print here, creates a cycle
-		fmt.Fprintf(ret, "\tPrimary index: prolly tree\n")
+		printWithIndendationLevel(level, ret, "\tAutoinc: %d\n", msg.AutoIncrementValue())
 
-		fmt.Fprintf(ret, "\tSecondary indexes: {\n\t%s\n", SerialMessage(msg.SecondaryIndexesBytes()).HumanReadableString())
-		fmt.Fprintf(ret, "\t}\n")
-		fmt.Fprintf(ret, "}")
+		printWithIndendationLevel(level, ret, "\tPrimary index: #%s\n", hash.Of(msg.PrimaryIndexBytes()))
```

Comment:
I'm saving printing it inline for a follow-up PR. The reason why that wasn't originally ported from noms_show was because it created a nasty dependency loop.

And you actually can look it up, although I'm not exactly sure why (and it only displays as a byte array because pretty-printing it would require fixing the aforementioned dependency loop.) I suspect that the object store has an entry pointing directly at the nested SerialMessage. FlatBuffers are pretty flexible that way.